File: tools/native/ollama_stream_probe.cpp
// Minimal Windows-only Ollama streaming latency probe.
//
// Build from the repo root with MinGW:
//   g++ tools/native/ollama_stream_probe.cpp -O2 -lws2_32 -o tools/native/ollama_stream_probe.exe
//
// Run:
//   tools\native\ollama_stream_probe.exe gemma4:e2b-it-q4_K_M
//
// This measures native HTTP streaming overhead only. Ollama still performs the
// model inference in its own llama.cpp runtime.

#define WIN32_LEAN_AND_MEAN

#include <winsock2.h>
#include <ws2tcpip.h>

#include <chrono>
#include <iostream>
#include <sstream>
#include <string>

namespace {

std::string json_escape(const std::string& value) {
    std::string escaped;
    escaped.reserve(value.size() + 8);
    for (char ch : value) {
        if (ch == '\\' || ch == '"') {
            escaped.push_back('\\');
        }
        if (ch == '\n') {
            escaped += "\\n";
        } else {
            escaped.push_back(ch);
        }
    }
    return escaped;
}

long long elapsed_ms(std::chrono::steady_clock::time_point started) {
    return std::chrono::duration_cast<std::chrono::milliseconds>(
               std::chrono::steady_clock::now() - started)
        .count();
}

}  // namespace

int main(int argc, char** argv) {
    const std::string model = argc > 1 ? argv[1] : "gemma4:e2b-it-q4_K_M";
    const std::string prompt =
        argc > 2 ? argv[2]
                 : "Cliente: Quiero una mesa cerca de la ventana porque viene una "
                   "persona mayor. Respuesta telefonica breve:";

    WSADATA wsa_data;
    if (WSAStartup(MAKEWORD(2, 2), &wsa_data) != 0) {
        std::cerr << "WSAStartup failed\n";
        return 1;
    }

    SOCKET sock = socket(AF_INET, SOCK_STREAM, IPPROTO_TCP);
    if (sock == INVALID_SOCKET) {
        std::cerr << "socket failed\n";
        WSACleanup();
        return 1;
    }

    sockaddr_in address{};
    address.sin_family = AF_INET;
    address.sin_port = htons(11434);
    address.sin_addr.s_addr = inet_addr("127.0.0.1");

    const auto started = std::chrono::steady_clock::now();
    if (connect(sock, reinterpret_cast<sockaddr*>(&address), sizeof(address)) == SOCKET_ERROR) {
        std::cerr << "connect failed. Is Ollama running?\n";
        closesocket(sock);
        WSACleanup();
        return 1;
    }

    const std::string body =
        std::string("{") +
        "\"model\":\"" + json_escape(model) + "\"," +
        "\"stream\":true,\"think\":false,\"keep_alive\":\"30m\"," +
        "\"system\":\"Responde como agente telefonico de restaurante. Maximo 25 palabras.\"," +
        "\"prompt\":\"" + json_escape(prompt) + "\"," +
        "\"options\":{\"num_predict\":48,\"num_ctx\":768,\"temperature\":0.2,\"top_p\":0.85,\"top_k\":20}" +
        "}";

    std::ostringstream request;
    request << "POST /api/generate HTTP/1.1\r\n"
            << "Host: 127.0.0.1:11434\r\n"
            << "Content-Type: application/json\r\n"
            << "Content-Length: " << body.size() << "\r\n"
            << "Connection: close\r\n\r\n"
            << body;

    const std::string raw_request = request.str();
    send(sock, raw_request.data(), static_cast<int>(raw_request.size()), 0);

    char buffer[4096];
    bool first_response_seen = false;
    std::string response;
    while (true) {
        int received = recv(sock, buffer, sizeof(buffer), 0);
        if (received <= 0) {
            break;
        }
        response.append(buffer, buffer + received);
        if (!first_response_seen && response.find("\"response\"") != std::string::npos) {
            first_response_seen = true;
            std::cout << "first_response_marker_ms=" << elapsed_ms(started) << "\n";
        }
        if (response.find(".") != std::string::npos && first_response_seen) {
            std::cout << "first_period_seen_ms=" << elapsed_ms(started) << "\n";
            break;
        }
    }

    std::cout << "total_probe_ms=" << elapsed_ms(started) << "\n";
    closesocket(sock);
    WSACleanup();
    return 0;
}

File: tools/native/ollama_stream_probe.rs
// Minimal Ollama streaming latency probe in Rust.
//
// Build from the repo root:
//   C:\Users\SERGI\.cargo\bin\rustc.exe tools/native/ollama_stream_probe.rs -O -o tools/native/ollama_stream_probe_rust.exe
//
// Run:
//   tools\native\ollama_stream_probe_rust.exe gemma4:e2b-it-q4_K_M
//   tools\native\ollama_stream_probe_rust.exe gemma4:e2b-it-q4_K_M "Cliente: ..." --num-thread 4 --num-ctx 256 --num-predict 24
//
// This uses only Rust stdlib. Ollama still performs inference in its own runtime.

use std::env;
use std::io::{Read, Write};
use std::net::TcpStream;
use std::time::Instant;

fn json_escape(value: &str) -> String {
    value
        .replace('\\', "\\\\")
        .replace('"', "\\\"")
        .replace('\n', "\\n")
}

fn elapsed_ms(started: Instant) -> u128 {
    started.elapsed().as_millis()
}

fn flag_value<'a>(args: &'a [String], flag: &str) -> Option<&'a str> {
    args.iter()
        .position(|arg| arg == flag)
        .and_then(|index| args.get(index + 1))
        .map(String::as_str)
}

fn parse_u32_flag(args: &[String], flag: &str, default_value: u32) -> u32 {
    flag_value(args, flag)
        .and_then(|value| value.parse::<u32>().ok())
        .unwrap_or(default_value)
}

fn extract_response_tokens(buffer: &str, consumed_lines: &mut usize) -> Vec<String> {
    let lines: Vec<&str> = buffer.lines().collect();
    let mut tokens = Vec::new();
    for line in lines.iter().skip(*consumed_lines) {
        if let Some(token) = extract_response_token(line) {
            tokens.push(token);
        }
    }
    *consumed_lines = lines.len();
    tokens
}

fn extract_response_token(line: &str) -> Option<String> {
    let marker = "\"response\":\"";
    let start = line.find(marker)? + marker.len();
    let mut escaped = false;
    let mut token = String::new();
    for character in line[start..].chars() {
        if escaped {
            token.push(match character {
                'n' => '\n',
                'r' => '\r',
                't' => '\t',
                '"' => '"',
                '\\' => '\\',
                other => other,
            });
            escaped = false;
            continue;
        }
        if character == '\\' {
            escaped = true;
            continue;
        }
        if character == '"' {
            return Some(token);
        }
        token.push(character);
    }
    None
}

fn main() -> std::io::Result<()> {
    let args: Vec<String> = env::args().collect();
    let model = args
        .get(1)
        .map(String::as_str)
        .unwrap_or("gemma4:e2b-it-q4_K_M");
    let prompt = args.get(2).map(String::as_str).unwrap_or(
        "Cliente: Quiero una mesa cerca de la ventana porque viene una persona mayor. Respuesta telefonica breve:",
    );
    let num_thread = parse_u32_flag(&args, "--num-thread", 4);
    let num_ctx = parse_u32_flag(&args, "--num-ctx", 256);
    let num_predict = parse_u32_flag(&args, "--num-predict", 24);

    let started = Instant::now();
    let mut stream = TcpStream::connect("127.0.0.1:11434")?;
    let body = format!(
        concat!(
            "{{",
            "\"model\":\"{}\",",
            "\"stream\":true,",
            "\"think\":false,",
            "\"keep_alive\":\"30m\",",
            "\"system\":\"Responde como agente telefonico de restaurante en Espana. Maximo 25 palabras.\",",
            "\"prompt\":\"{}\",",
            "\"options\":{{\"num_thread\":{},\"num_predict\":{},\"num_ctx\":{},\"temperature\":0.1,\"top_p\":0.8,\"top_k\":20,\"stop\":[\"\\n\"]}}",
            "}}"
        ),
        json_escape(model),
        json_escape(prompt),
        num_thread,
        num_predict,
        num_ctx
    );
    let request = format!(
        concat!(
            "POST /api/generate HTTP/1.1\r\n",
            "Host: 127.0.0.1:11434\r\n",
            "Content-Type: application/json\r\n",
            "Content-Length: {}\r\n",
            "Connection: close\r\n\r\n",
            "{}"
        ),
        body.len(),
        body
    );
    stream.write_all(request.as_bytes())?;

    let mut buffer = [0_u8; 4096];
    let mut response = String::new();
    let mut collected = String::new();
    let mut first_response_seen = false;
    let mut consumed_lines = 0_usize;
    println!(
        "settings model={} num_thread={} num_ctx={} num_predict={}",
        model, num_thread, num_ctx, num_predict
    );
    loop {
        let received = stream.read(&mut buffer)?;
        if received == 0 {
            break;
        }
        response.push_str(&String::from_utf8_lossy(&buffer[..received]));
        for token in extract_response_tokens(&response, &mut consumed_lines) {
            if !first_response_seen {
                first_response_seen = true;
                println!("first_response_marker_ms={}", elapsed_ms(started));
            }
            collected.push_str(&token);
        }
        if first_response_seen && collected.contains('.') {
            println!("first_period_seen_ms={}", elapsed_ms(started));
            break;
        }
    }
    println!("total_probe_ms={}", elapsed_ms(started));
    println!("collected={}", collected.trim());
    Ok(())
}
